libraries/lib-snapping/ProjectSnap.h

```diff
@@ -0,0 +1,51 @@
+/*  SPDX-License-Identifier: GPL-2.0-or-later */
+/*!********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  ProjectSnap.h
+
+  Dmitry Vedenko
+
+**********************************************************************/
+#pragma once
+
+#include "ClientData.h"
+#include "Observer.h"
+#include "SnapUtils.h"
+
+class AudacityProject;
+
+struct SnapChangedMessage final
+{
+   SnapMode   newSnapMode { SnapMode::SNAP_OFF };
+   Identifier newSnapTo { "seconds" };
+};
+
+///\brief Holds project sample rate
+class SNAPPING_API ProjectSnap final :
+    public ClientData::Base,
+    public Observer::Publisher<SnapChangedMessage>
+{
+public:
+   static ProjectSnap& Get(AudacityProject& project);
+   static const ProjectSnap& Get(const AudacityProject& project);
+
+   explicit ProjectSnap(const AudacityProject& project);
+   ProjectSnap(const ProjectSnap&) = delete;
+   ProjectSnap& operator=(const ProjectSnap&) = delete;
```

Comment:
Did I ever explain the PROHIBITED macro and why I don't still use it and sometimes remove it?

Wow, seven years ago. 9e5d31d347873f7e54f5a4b7e8528fa61490b792

While purging all the old naked news and deletes, I wanted easy text scans to see where remaining ones were that needed work.  `safenew` also dates from this time for certain news that are correctly unpaired and don't need smart pointers -- any construction of a wxWindow with a non-null parent, because the parent assumes deletion responsibility (and I think Qt works the same?).